File: projections/dotnet/foundation/hlp.ui.tone-style/ToneStyle.cs
using Harborline.Foundation.Builder;

namespace Harborline.Foundation.Theming;

/// <summary>Authored semantic CSS custom-property references for a builder lens tone.</summary>
public sealed record ToneStyle(string Swatch, string Border, string SoftBackground, string Text);

/// <summary>Projection-neutral mapping from the closed lens-tone vocabulary to theme tokens.</summary>
public static class ToneStyles
{
    /// <summary>Resolves one tone without consulting host state.</summary>
    public static ToneStyle Resolve(LensTone tone) => tone switch
    {
        LensTone.Accent or LensTone.SensitivityLow => Accent,
        LensTone.Warning or LensTone.SensitivityMedium => Warning,
        LensTone.Danger or LensTone.SensitivityHigh => Danger,
        LensTone.Success => Success,
        LensTone.Info => Info,
        LensTone.Muted or LensTone.SensitivityNone => Muted,
        _ => throw new ArgumentOutOfRangeException(nameof(tone), tone, "Unknown lens tone."),
    };

    private static readonly ToneStyle Accent = new(
        "var(--color-accent)", "var(--color-accent)",
        "var(--color-accent-soft)", "var(--color-accent-foreground)");
    private static readonly ToneStyle Warning = new(
        "var(--color-warning)", "var(--color-warning)",
        "var(--color-warning-soft)", "var(--color-warning)");
    private static readonly ToneStyle Danger = new(
        "var(--color-danger)", "var(--color-danger)",
        "var(--color-danger-soft)", "var(--color-danger)");
    private static readonly ToneStyle Success = new(
        "var(--color-success)", "var(--color-success)",
        "var(--color-success-soft)", "var(--color-success)");
    private static readonly ToneStyle Info = new(
        "var(--color-secondary)", "var(--color-secondary)",
        "var(--color-muted)", "var(--color-secondary-foreground)");
    private static readonly ToneStyle Muted = new(
        "var(--color-muted-foreground)", "var(--color-border)",
        "var(--color-muted)", "var(--color-muted-foreground)");
}

File: projections/react/ui/hlp.ui.tone-style/src/toneStyle.ts
import type { LensTone } from '@harborline-platform/hlp.ui.aspect-lens'

export interface ToneStyle {
  swatch: string
  border: string
  softBg: string
  text: string
}

const styles: Readonly<Record<LensTone, Readonly<ToneStyle>>> = {
  accent: {
    swatch: 'var(--color-accent)',
    border: 'var(--color-accent)',
    softBg: 'var(--color-accent-soft)',
    text: 'var(--color-accent-foreground)',
  },
  warning: {
    swatch: 'var(--color-warning)',
    border: 'var(--color-warning)',
    softBg: 'var(--color-warning-soft)',
    text: 'var(--color-warning)',
  },
  danger: {
    swatch: 'var(--color-danger)',
    border: 'var(--color-danger)',
    softBg: 'var(--color-danger-soft)',
    text: 'var(--color-danger)',
  },
  success: {
    swatch: 'var(--color-success)',
    border: 'var(--color-success)',
    softBg: 'var(--color-success-soft)',
    text: 'var(--color-success)',
  },
  info: {
    swatch: 'var(--color-secondary)',
    border: 'var(--color-secondary)',
    softBg: 'var(--color-muted)',
    text: 'var(--color-secondary-foreground)',
  },
  muted: {
    swatch: 'var(--color-muted-foreground)',
    border: 'var(--color-border)',
    softBg: 'var(--color-muted)',
    text: 'var(--color-muted-foreground)',
  },
  'sensitivity-none': {
    swatch: 'var(--color-muted-foreground)',
    border: 'var(--color-border)',
    softBg: 'var(--color-muted)',
    text: 'var(--color-muted-foreground)',
  },
  'sensitivity-low': {
    swatch: 'var(--color-accent)',
    border: 'var(--color-accent)',
    softBg: 'var(--color-accent-soft)',
    text: 'var(--color-accent-foreground)',
  },
  'sensitivity-medium': {
    swatch: 'var(--color-warning)',
    border: 'var(--color-warning)',
    softBg: 'var(--color-warning-soft)',
    text: 'var(--color-warning)',
  },
  'sensitivity-high': {
    swatch: 'var(--color-danger)',
    border: 'var(--color-danger)',
    softBg: 'var(--color-danger-soft)',
    text: 'var(--color-danger)',
  },
}

export function toneStyle(tone: LensTone): ToneStyle {
  return styles[tone]
}
